lib/JSON/Validator.pm

```diff
@@ -258,7 +258,10 @@ sub _resolve_schema {
 
         # int($v) returns the memory address of the reference. The "if" below
         # will not resolve the same "$ref" over again.
-        next if ref $v and $self->{seen}{int($v)}++;
+        next if ref $v and $self->{seen}{int($v)};
```

Comment:
This solution seems to fix my original issue and at the moment I don't have anything against this change.